### PR TITLE
Encode the datetime values in the bundle.Link.URL in the resource #3245

### PR DESCRIPTION
**Pull request by prb112**
Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-search/src/main/java/com/ibm/fhir/search/uri/UriBuilder.java [hunk 1]
```diff
@@ -100,7 +100,10 @@ public String toSearchSelfUri() throws URISyntaxException {
         URI selfUri = new URI(requestUri.getScheme(), requestUri.getAuthority(), requestUri.getPath(),
                 queryString.toString(), null);
 
-        return selfUri.toString();
+        // ... originally the approach was to modify the Type.Date parameter value, however that's unreliable
+        // as a '+' is a control character which the URI code then tries to escape the escaped value.
+        String urlString = selfUri.toString();
+        return urlString.replace("+", "%2B");
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/uri/UriBuilder.java:106]: Originally I tried manipulating the Type.Date on 294's stream. This turns out that java.net.URI has an encoding for '%' which causes it to get encoded so you end up with `%252B` which is not ideal. There are alternatives such as switching the framework to use Guava and URLEscapers, however that is a heavy lift.  This case it converts the + to be a plus and the framework on read doesn't convert to SPACE

**Review by lmsurpre — APPROVED** [review]
LGTM

**Review by tbieste — COMMENTED** [review]
I looked at the ways the self/next/prev links are built in FHIRRestHelper.java, and I'm not sure if the prev link would see this change.  Perhaps a ServerTest (FVT) level test to ensure it's correct.

**Comment by prb112** [comment]
> I looked at the ways the self/next/prev links are built in FHIRRestHelper.java, and I'm not sure if the prev link would see this change. Perhaps a ServerTest (FVT) level test to ensure it's correct.

FHIRRestHelper uses the selfUrl from SearchUtil to build the links, and then uses the selfUrl with the fix to build the prev and next links. I'll add an integration test.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java [hunk 2]
```diff
@@ -576,6 +578,24 @@ public void testSearchAllUsingOneTag() throws Exception {
         }
     }
 
+    @Test(groups = { "server-search-all" }, dependsOnMethods = { "testSearchAllUsingOneTag" })
+    public void testSearchAllUsingOneTagForPrevNext() throws Exception {
```
Comment (prb112) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchAllTest.java:582]: @tbieste here is the FVT

**Review by tbieste — APPROVED** [review]

**Pull request merged**